### PR TITLE
 add basic rotation

### PR DESCRIPTION
**Pull request by deads2k**
This adds certificate rotation for the aggregation proxy client cert/key pair, the signing cert/key pair for it, and the ca-bundle configmap as well.

@sttts @mfojtik 
@openshift/sig-auth I'm not very opinionated about exactly how it works, but this is roughly what we want to achieve.  A re-entrant controller loop that can manage a signer, a trust bundle, and a client or serving cert.  It also works properly with multiple parties using the same signer or ca bundle.  I'd like to start with this and I'm happy to switch to something more robust/better later

### CODE REVIEW

**Comment by ericavonb** [comment]
@deads2k why are we doing this in here? we have rotation functionality, in the service-serving-cert-signer (currently in progress/review).

We're trying to move certificate management to a centralized, dedicated place where we can better manage all the assets and further modernize the architecture. No new certificate logic should be added to components themselves.

/hold

**Comment by deads2k** [comment]
discussed in bluejeans, this is ok for @mrogers950 to review.

/hold cancel

File: pkg/operator/aggregatorproxyrotation/target.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	targetCertKeyPairSecret := originalTargetCertKeyPairSecret.DeepCopy()
+	if apierrors.IsNotFound(err) {
+		// create an empty one
+		targetCertKeyPairSecret = &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: c.targetNamespace, Name: c.targetCertKeyPairSecretName}}
+	}
+	if needNewTargetCertKeyPair(targetCertKeyPairSecret.Annotations, signingCertKeyPair, c.targetCertKeyPairValidity, c.newTargetPercentage) {
+		c.eventRecorder.Eventf("TargetUpdateRequired", "%q in %q requires a new signing cert/key pair", c.targetCertKeyPairSecretName, c.targetNamespace)
```
Comment (mrogers950) [inline pkg/operator/aggregatorproxyrotation/target.go]: s/signing/target

File: pkg/operator/aggregatorproxyrotation/target.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	if signerCommonName != signer.Config.Certs[0].Subject.CommonName {
+		return true
+	}
+
+	return false
+}
+
+// setTargetCertKeyPairSecret creates a new cert/key pair and sets them in the secret
+func setTargetCertKeyPairSecret(targetCertKeyPairSecret *corev1.Secret, validity time.Duration, signer *crypto.CA, user user.Info, servingHostnames []string, servingCertificateExtensionFn crypto.CertificateExtensionFunc) error {
```
Comment (mrogers950) [inline pkg/operator/aggregatorproxyrotation/target.go]: Can this be split into server/client setters, perhaps as a func passed to the controller? It confused me that it did one or the other until I saw it was shared between the client and server update controllers.

**Comment by deads2k** [comment]
nits addressed.  Now with unit tests.  If we like this shape, we can merge it here and I'll do a straight move to library-go for later re-use.

**Comment by deads2k** [comment]
> How do we relate to cert renewal for the router? Do they rotate on their own by duplicating this?

This is actually built to be flexible and re-useable.  I may end up providing three different listers/informers, but in concept you provide your three API objects and this will produce the output you need.  We'll be trying it out for three or four in the master team operators.

File: pkg/operator/aggregatorproxyrotation/client_cert_rotation_controller.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+	return true
+}
+
+// eventHandler queues the operator to check spec and status
+func (c *CertRotationController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/operator/aggregatorproxyrotation/client_cert_rotation_controller.go]: oh, right. reminds me that the sync loops should check and don't act on deleted objects

File: pkg/operator/aggregatorproxyrotation/target.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	if targetCertKeyPairSecret.Data == nil {
+		targetCertKeyPairSecret.Data = map[string][]byte{}
+	}
+
+	if len(servingHostnames) > 0 {
+		servingCertKeyPair, err := signer.MakeServerCertForDuration(sets.NewString(servingHostnames...), validity, servingCertificateExtensionFn)
+		if err != nil {
+			return err
+		}
+		targetCertKeyPairSecret.Data["tls.crt"], targetCertKeyPairSecret.Data["tls.key"], err = servingCertKeyPair.GetPEMBytes()
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/operator/aggregatorproxyrotation/target.go]: point of constants isn't shortening, but maintainability and avoiding typos. these constants are part of our API even https://github.com/kubernetes/kubernetes/blob/6ec1ebd6ce65b7b05f195551e89649428d8c2738/pkg/apis/core/types.go#L4439

**Pull request merged**